DOI.js

```diff
@@ -93,38 +96,50 @@ function completeDOIs(doc) {
 	}
 }
 
-function retrieveDOIs(DOIs, doc) {
+function retrieveDOIs(DOIs, doc, provider) {
 	__num_DOIs = DOIs.length;
 
 	for(var i=0, n=DOIs.length; i<n; i++) {
 		(function(doc, DOI) {
 			var translate = Zotero.loadTranslator("search");
+			//Default = CrossRef
 			translate.setTranslator("11645bd1-0420-45c1-badb-53fb41eeb753");
-	
+			if (provider=="DataCite") {
+				translate.setTranslator("9f1fb86b-92c8-4db7-b8ee-0b481d456428");
+			}
 			var item = {"itemType":"journalArticle", "DOI":DOI};
 			translate.setSearch(item);
 	
 			// don't save when item is done
 			translate.setHandler("itemDone", function(translate, item) {
 				if (!item.title) {
-					Zotero.debug("No title available for " + DOI);
+					Zotero.debug("No title available for " + item.DOI);
+					item.title = item.DOI;
 					return;
 				}
 				
-				item.repository = "CrossRef";
-				items[DOI] = item;
-				selectArray[DOI] = item.title;
+				items[item.DOI] = item;
+				selectArray[item.DOI] = item.title;
 			});
 	
 			translate.setHandler("done", function(translate) {
 				__num_DOIs--;
 				if(__num_DOIs <= 0) {
-					completeDOIs(doc);
+					Z.debug("after");
+					if (firstEncounter) {
+						firstEncounter = false;//don't loop forever
+						retrieveDOIs(remainingDOIs, doc, "DataCite");
+					} else {
+						completeDOIs(doc);
+					}
 				}
 			});
 	
 			// Don't throw on error
-			translate.setHandler("error", function() {});
+			translate.setHandler("error", function() {
+				Z.debug("err with DOI: " + DOI);
+				remainingDOIs.push(DOI);
```

Comment:
I guess we have to switch the creation of the `remainingDOIs` then. We can copy all `DOIs` first into that array and delete the ones which are successful. This can be done in the itemDone-handler where we have access to the current `item`.